### PR TITLE
Colin/agents-threaded-event-reader

### LINKED ISSUE #153: Multithreaded event handling for agents
This issue came up in thinking about how to allow agents to asynchronously await events in order to dictate their actions. There are likely multiple ways to accomplish this, but here's an outline.

- [ ] Allow the `SimulationEnvironment` to own write control over a buffer where events are written. Upon any call to execution, events should be echoed into this buffer.
- [ ] Agents need read only access to the buffer and as such they can read from this buffer without the buffer needing to be mutex since agents will not write -- only the environment/EVM instance will write. Agents need a function to access these logs.
- [ ] Agents will also need some type of async await function that lets them constantly poll for updates in event logs.

### PR DESCRIPTION
**Pull request by Autoparallel**
Attempting to close #153. 

Still need to build the functionality of actually awaiting new events. Trying to model this after ethers.

### CODE REVIEW

**Review by 0xJepsen — COMMENTED** [review]
This looks good to me other than the lint failing, what other things need to be done on this PR?

**Comment by Autoparallel** [comment]
> This looks good to me other than the lint failing, what other things need to be done on this PR?

It was not very clean at that point -- just working. I cleaned it up and it's looking good now. Opening PR now.

**Comment by Autoparallel** [comment]
Closes #153

**Comment by Autoparallel** [comment]
I need to set up dispatching so messages aren't consumed on other threads. Working on this now.

**Comment by Autoparallel** [comment]
I think this should close it. See the reworked test.

File: crates/simulate/src/lib.rs [hunk 1]
```diff
@@ -316,6 +316,57 @@ mod tests {
             }
         });
 
+        let reader_for_admin = manager.agents.get("admin").unwrap().receiver();
+        let writer_base_contract_for_admin = writer.base_contract.clone();
+        let admin_handle = thread::spawn(move || {
+            let mut i = 0;
+            while let Ok(logs) = reader_for_admin.recv() {
```
Comment (0xJepsen) [inline crates/simulate/src/lib.rs:323]: nice

**Review by 0xJepsen — APPROVED** [review]
lgtm!

**Pull request merged**